src/client/components/DicePanel.tsx

```diff
@@ -1,57 +1,382 @@
 } from "@fortawesome/free-solid-svg-icons";
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 import clsx from "clsx";
-import React, { useState } from "react";
+import React, { useState, useEffect } from "react";
+import {
+  playerAddDiceTemplateCategory,
+  playerUpdateDiceTemplateCategory,
+  playerDeleteDiceTemplateCategory,
+} from "../../shared/actions";
+import { DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME } from "../../shared/constants";
+import {
+  categoryIcons,
+  characterStatNames,
+  fixedCategoryIcons,
+  iconMap,
+  RRDiceTemplatePart,
+  RRDiceTemplatePartDice,
+  skillMap,
+  skillNames,
+  userCategoryIcons,
+} from "../../shared/state";
+import { rrid } from "../../shared/util";
+import {
+  RRDiceTemplate,
+  RRDiceTemplateCategory,
+} from "../../shared/validation";
+import { useMyProps } from "../myself";
+import { useServerDispatch, useServerState } from "../state";
 import { DiceInput } from "./DiceInput";
 import { DiceInterface } from "./DiceInterface";
-import { DiceTemplates } from "./DiceTemplates";
-
-export const diceCategories = [
-  faFireAlt,
-  faBalanceScale,
-  faShieldAlt,
-  faMagic,
-  faDragon,
-  faBroom,
-];
+import { DiceTemplates, GeneratedDiceTemplates } from "./DiceTemplates";
+import { Popover } from "./Popover";
+import { Button } from "./ui/Button";
+import { SmartTextInput } from "./ui/TextInput";
 
 export const DicePanel = React.memo(function DicePanel() {
-  const [active, setActive] = useState(0);
+  const [active, setActive] = useState("Input");
+  const myself = useMyProps("diceTemplateCategories", "id", "mainCharacterId");
+  const dispatch = useServerDispatch();
+
+  useEffect(() => {
+    if (
+      active !== "Skills" &&
+      active !== "STs" &&
+      myself.diceTemplateCategories.find((cat) => cat.id === active) ===
+        undefined
+    ) {
+      setActive("Input");
+    }
+  }, [myself.diceTemplateCategories, active]);
+
+  function addTemplateCategory() {
+    const freeIcon = userCategoryIcons.find((userIcon) => {
+      return !myself.diceTemplateCategories
+        .map(({ icon }: { icon: typeof categoryIcons[number] }) => icon)
+        .includes(userIcon);
+    });
+
+    if (!freeIcon) return;
+
+    dispatch({
+      actions: [
+        playerAddDiceTemplateCategory({
+          id: myself.id,
+          category: {
+            id: rrid<RRDiceTemplateCategory>(),
+            icon: freeIcon,
+            categoryName: "",
+            templates: [],
+          },
+        }),
+      ],
+      optimisticKey: "diceTemplateCategories",
+      syncToServerThrottle: DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME,
+    });
+  }
+
+  const d20Part: RRDiceTemplatePartDice = {
+    id: rrid<RRDiceTemplatePart>(),
+    type: "dice",
+    faces: 20,
+    count: 1,
+    negated: false,
+    modified: "none",
+    damage: { type: null },
+  };
+  const characters = useServerState((state) => state.characters);
+  const character = characters.entities[myself.mainCharacterId!];
```

Comment:
Can we move the selection of the main character into the `useServerState` selector, so that we don't unnecessarily subscribe to all character changes? Also, `myself.mainCharacterId!` seems fishy to me - what if I don't have a main character set?